### PR TITLE
Mason v0.1.0: Chapel's Package Manager

### PR DESCRIPTION
**Pull request by Spartee**
This pull request is the first for the Mason Package Manager. This PR includes the mason command line interface with the following commands:

Installed Mason Commands:
      new
      update
      build
      run 
      clean
      help
      version

An example usage is as follows:

- mason new [ options ] ( project name )
- mason run --build

OR 

- mason new [ options ] ( project name ) 
- mason update
- mason build [ options ]
- mason run [ options ]           -- including build after run compiles the program as well. 

TO DO 
- [x] Unit tests
- [x] Mason help
- [x] Makefile
- [x] optimizations and target/release
- [x] reduce memory leakage
- [x] IVRS version number bug

### CODE REVIEW

File: tools/mason/MasonBuild.chpl [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+  for dir in listdir(repos) {
+    if dir == dependency {
+      exists = true;
+    }
+  }
+  return exists;
+}
+
+
+proc runCommand(cmd) {
```
Comment (ben-albrecht) [inline tools/mason/MasonBuild.chpl]: Perhaps this should be put into a `MasonUtils.chpl` module to be used within other `mason` modules?

File: tools/mason/MasonNew.chpl [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+
+proc moveFile(name: string, file: string) {
+  var dir = realPath(name);
+  var command = "mv " + file + ' ' + dir;
+  var result = runCommand(command);
+}
+
+
+proc runCommand(command): int {
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl]: `MasonUtils.chpl` - (see previous comment)

File: tools/mason/mason.chpl [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    if args[2] == "build" {
+      masonBuild(args);
+    }
+  }
+  if isFile("Mason.lock") {
+    var command = "target/debug/" + toRun;
+    runCommand(command);
+  }
+  else {
+    writeln("call mason run from the top level of your projects directory");
```
Comment (ben-albrecht) [inline tools/mason/mason.chpl]: We should remove this constraint at some point.

File: tools/mason/MasonNew.chpl [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+  }
+  else {
+    makeBasicToml(name);
+    makeProjectFiles(name);
+    writeln("Mason initialized your new project: " + name);
+  }
+}
+
+
+proc gitInit(name: string) : int {
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl]: I'd be tempted to wrap all of the `git` command functions into a helper-module, e.g. `Git.chpl` -- this would make switching over to `libgit` as a dependency easier in the future.

File: tools/mason/MasonNew.chpl [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+proc makeProjectFiles(name: string) {
+  var makeSrc = runCommand("mkdir src");
+  var makeLib = runCommand("touch src/" + name + '.chpl');
+  moveFile(name, "src");
+  moveFile(name, "Mason.toml");
+}
+
+
+proc moveFile(name: string, file: string) {
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl]: Note that `file` is an IO type, so it's preferable to not use it as a variable name, if possible. Maybe `src` and `dest` works better here?

File: tools/mason/MasonBuild.chpl [hunk 6]
```diff
@@ -0,0 +1,147 @@
+
+use TOML;
+use Spawn;
+use FileSystem;
+use mason;
```
Comment (ben-albrecht) [inline tools/mason/MasonBuild.chpl]: Ideally, nothing else should `use mason`

File: tools/mason/MasonBuild.chpl [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+  }
+  else {
+    writeln("Cannot build: no Mason.lock found");
+  }
+} 
+
+
+/* Creates the rest of the project structure */
+proc makeTargetFiles() {
+  var makeTarget = "mkdir target";
```
Comment (ben-albrecht) [inline tools/mason/MasonBuild.chpl]: We can use `mkdir -p` and omit the checks for `ifDirExists`

File: tools/mason/mason.chpl [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+  when '--version' do printVersion();
+  otherwise {
+    writeln('error: no such subcommand');
+    writeln('try mason --help');
+    exit();}
+  }
+}
+
+
+proc masonNew(args) {
```
Comment (ben-albrecht) [inline tools/mason/mason.chpl]: This can live in `MasonNew.chpl`

File: tools/mason/mason.chpl [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    otherwise {
+      InitProject(args);
+    }
+    }
+  }
+}
+
+
+
+proc masonBuild(args) {
```
Comment (ben-albrecht) [inline tools/mason/mason.chpl]: and this in `MasonBuild.chpl`, etc.

File: tools/mason/mason.chpl [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    writeln("call mason run from the top level of your projects directory");
+  }
+}
+
+// TODO
+proc masonInit(args) {}
+proc masonClean(args) {}
+proc masonDoc(args) {}
+
+proc runCommand(cmd) {
```
Comment (ben-albrecht) [inline tools/mason/mason.chpl]: `MasonUtils.chpl`

File: tools/mason/mason.chpl [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+   var splitCmd = cmd.split();
+   var process = spawn(splitCmd, stdout=PIPE);
+   process.wait();
+   
+   for line in process.stdout.lines() {
+     writeln(line);
+   }
+}
+
+/* Gets envoirment variables for spawn commands */
```
Comment (ben-albrecht) [inline tools/mason/mason.chpl]: typo - try adding these files to `chplspell` and run it on them.

File: tools/mason/mason.chpl [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+proc masonInit(args) {}
+proc masonClean(args) {}
+proc masonDoc(args) {}
+
+proc runCommand(cmd) {
+   var splitCmd = cmd.split();
+   var process = spawn(splitCmd, stdout=PIPE);
+   process.wait();
+   
+   for line in process.stdout.lines() {
```
Comment (ben-albrecht) [inline tools/mason/mason.chpl]: Do we *always* want to print the command output? Perhaps this should be an option in the `runCommand()` args.

File: tools/mason/mason.chpl [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+  writeln('    --list              List installed commands');
+  writeln();
+  writeln('Some common mason commands are (see all commands with --list):');
+  writeln('    new         Create a new mason project');
+  writeln('    update      Update/Generate Mason.lock');
+  writeln('    build       Compile the current project');
+  //writeln('    clean       Remove the target directory');
+  //writeln('    doc         Build this project\'s and its dependencies\' documentation');
+  //writeln('    init        Create a new mason project in an existing directory');
+  writeln('    run         Build and execute src/main.chpl');
```
Comment (ben-albrecht) [inline tools/mason/mason.chpl]: Is it always `main.chpl`?

File: tools/mason/mason.chpl [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+use TOML;
+use MasonNew;
+use MasonBuild;
+use MasonUpdate;
+
+
+proc main(args: [] string) {
+  if args.size < 2 {
+    masonHelp();
+    exit();}
```
Comment (ben-albrecht) [inline tools/mason/mason.chpl]: style nit-pick:

```chpl
  exit();
}
```

File: tools/mason/mason.chpl [hunk 15]
```diff
@@ -0,0 +1,152 @@
+
+use FileSystem;
+use Help;
+use Spawn;
+use TOML;
```
Comment (ben-albrecht) [inline tools/mason/mason.chpl]: Be sure to update the `use` statements here, once commands are distributed into other modules. e.g. Do we still need `Spawn` and `TOML` here?

File: tools/mason/MasonNew.chpl (hunk 2 above)
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl]: As noted elsewhere, you'll probably want to differentiate the `runCommand` behaviors with some arguments, overloads, or different names.

File: tools/mason/MasonNew.chpl (hunk 5 above)
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl]: Should we check if files exist before moving? What happens if it does?

File: tools/mason/MasonNew.chpl [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+  var tomlFile = open("Mason.toml", iomode.cw);
+  var tomlWriter = tomlFile.writer();
+  tomlWriter.write(baseToml);
+  tomlWriter.close();
+}
+  
+
+proc makeProjectFiles(name: string) {
+  var makeSrc = runCommand("mkdir src");
+  var makeLib = runCommand("touch src/" + name + '.chpl');
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl]: TODO: Make the created file have some standard template, e.g.

```chpl
/* name module */
module name {

}
```

**Comment by ben-albrecht** [comment]
We'll want to add some unit tests and functional tests as well.

File: tools/mason/mason.chpl (hunk 12 above)
Comment (Spartee) [inline tools/mason/mason.chpl]: So I ended up going with two different versions in MasonUtils. One that returns status of command and one that does not. refer to -> [https://github.com/chapel-lang/chapel/pull/6951/commits/f1880b0067d835020124baddd9a36d27d8e0ec02#diff-71a91c1fdf3c3d0e5a204b343fafd8d6](url)

File: tools/mason/MasonBuild.chpl (hunk 7 above)
Comment (ben-albrecht) [inline tools/mason/MasonBuild.chpl]: Actually, we can use: [FileSystem.mkdir](http://chapel.cray.com/docs/master/modules/standard/FileSystem.html#FileSystem.mkdir)

File: tools/mason/MasonUtils.chpl [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+  sub.wait();
+  return sub.exit_status;
+}
+
+
+/* Moves a file from curdir to a folder in
+   curdir */
+proc moveFile(name: string, src: string) {
+  var dir = realPath(name);
+  var command = "mv " + src + ' ' + dir;
```
Comment (ben-albrecht) [inline tools/mason/MasonUtils.chpl]: `FileSystem.rename()` (for files)
`FileSystem.moveDir()` (for directories)

File: tools/mason/MasonBuild.chpl (hunk 7 above)
Comment (Spartee) [inline tools/mason/MasonBuild.chpl]: Would it not be faster to use one conditional isDir(target) instead of spawning a number of "mkdir -p <folder>" commands?

File: tools/mason/MasonNew.chpl [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+  var command = "git init -q " + name;
+  return runWithStatus(command);
+}
+
+proc addGitIgnore(name: string) {
+  var toIgnore = "\ntarget/\nMason.lock";
+  var gitIgnore = open(name+"/.gitignore", iomode.cw);
+  var GIwriter = gitIgnore.writer();
+  GIwriter.write(toIgnore);
+  GIwriter.close();
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl]: `touch` could be accomplished by opening the file for writing:

```chpl
proc touch(f) {
  var filehandle = open(f, iomode.cw);
  filehandle.close();
}
```

File: tools/mason/MasonBuild.chpl (hunk 7 above)
Comment (ben-albrecht) [inline tools/mason/MasonBuild.chpl]: I don't think performance is crucial in this function, and using the library calls would be much cleaner.

File: tools/mason/Makefile [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+REGISTRY=https://github.com/Spartee/mason-registry
+
+all: mason
+
+mason:
+	@echo "Building Mason..."
+	#build
+	#install
+
+build:
```
Comment (ben-albrecht) [inline tools/mason/Makefile]: Let's make this `mason:`

File: tools/mason/Makefile [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+	@mkdir -p ~/.mason
+	@mkdir -p ~/.mason/src
+	@git -C ~/.mason/ clone $(REGISTRY) registry
+	@echo "Building Mason..."
+	@chpl mason.chpl
```
Comment (ben-albrecht) [inline tools/mason/Makefile]: Does this assume `TOML.chpl` is in `modules/packages`? Perhaps this should (temporarily) pass a `-M test/modules/packages/TOML` until then.

@benharsh aptly pointed out that we'll need to promote `TOML.chpl` to `modules/packages/` by the 1.16 release (9/18), because `test/` is not included in the release tarball.

File: doc/rst/tools/mason/mason.rst [hunk 21]
```diff
@@ -0,0 +1,339 @@
...
+Each versioned manifest file would be identical to the manifest file in the top-level directory
+of the package repository, with one additional field not required in the repository manifest,
+a URL pointing to the repository and revision in which the version is located.
+
+Continuing the example from before, the 'registry' ``Mason.toml`` would include the additional source field:
+
+.. code-block:: text
+
+     [brick]
+     name = "hello_world"
```
Comment (cassella) [inline doc/rst/tools/mason/mason.rst]: Is it intentional that the name ``hello_world`` here doesn't match the ``MyProject`` above?

I gather ``brick`` is a required section, as is ``root`` in the lock file below?

File: doc/rst/tools/mason/mason.rst [hunk 22]
```diff
@@ -0,0 +1,339 @@
...
+      1.2.0.toml
+   VisualDebug/
+      2.2.0.toml
+      2.2.1.toml
+
+Each versioned manifest file would be identical to the manifest file in the top-level directory
+of the package repository, with one additional field not required in the repository manifest,
+a URL pointing to the repository and revision in which the version is located.
+
+Continuing the example from before, the 'registry' ``Mason.toml`` would include the additional source field:
```
Comment (cassella) [inline doc/rst/tools/mason/mason.rst]: This would actually be ``0.1.0.toml``, not ``Mason.toml``, right?

File: doc/rst/tools/mason/mason.rst [hunk 23]
```diff
@@ -0,0 +1,339 @@
...
+
+
+
+The Lock File
+=============
+
+The lock file ``Mason.lock`` is generated after running a mason update command. The user should
+never manually edit the lock file as it is intended to "lock" in the settings of a certain 
+project build iteration. ``Mason.lock`` is added by default to the .gitignore when a new project 
+is created. If you intention is to create a binary application package that does not need to
```
Comment (cassella) [inline doc/rst/tools/mason/mason.rst]: s/you/your/

File: tools/mason/Makefile [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+link=$(bdir)/mason
+
+
+REGISTRY=https://github.com/Spartee/mason-registry
+
+all: mason install
+
+mason:
+	@mkdir -p ~/.mason
+	@mkdir -p ~/.mason/src
```
Comment (cassella) [inline tools/mason/Makefile]: FYI, ``mkdir -p ~/.mason/src`` will create the ``~/.mason`` directory if it doesn't already exist.

File: tools/mason/MasonUtils.chpl [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+  }
+  return exists;
+}
+
+
+proc getMasonHome(): string {
+  var masonHome = getEnv("CHPL_MASON_HOME");
+  if masonHome == '' {
+    var home = getEnv('HOME');
+    if isFile(home + '/.mason') {
```
Comment (Spartee) [inline tools/mason/MasonUtils.chpl]: Change to isDir().

File: tools/mason/Makefile [hunk 26]
```diff
@@ -0,0 +1,44 @@
...
+
+TOML=$(CHPL_HOME)/test/modules/packages/Toml/
+REGISTRY=https://github.com/chapel-lang/mason-registry
+
+all: mason install
+
+mason:
+	@mkdir -p $(CHPL_MASON_HOME)/.mason
+	@mkdir -p $(CHPL_MASON_HOME)/.mason/src
+	@git -C $(CHPL_MASON_HOME)/.mason/ clone $(REGISTRY) registry
```
Comment (ben-albrecht) [inline tools/mason/Makefile]: We should check if the registry is already cloned before doing this. If it is, you get:

```
fatal: destination path 'registry' already exists and is not an empty directory.
```

File: tools/mason/Makefile [hunk 27]
```diff
@@ -0,0 +1,44 @@
...
+ifneq ($(wildcard $(link)),)
+	@echo "Removing old symbolic link..."
+	rm -f $(link)
+	@echo
+endif
+	@echo "Installing symbolic link..."
+	mkdir -p $(bdir)
+	ln -s $(shell pwd)/mason $(link)
+
+clean:
```
Comment (ben-albrecht) [inline tools/mason/Makefile]: Should `make clean` remove the registry? If not, maybe `make clobber` should?

File: tools/mason/Makefile [hunk 28]
```diff
@@ -0,0 +1,44 @@
...
+all: mason install
+
+mason:
+	@mkdir -p $(CHPL_MASON_HOME)/.mason
+	@mkdir -p $(CHPL_MASON_HOME)/.mason/src
+	@git -C $(CHPL_MASON_HOME)/.mason/ clone $(REGISTRY) registry
+	@echo "Building Mason..."
+	@chpl mason.chpl -M $(TOML)
+
+install:
```
Comment (ben-albrecht) [inline tools/mason/Makefile]: `install` should depend on `mason`, so that calling `make install` will both build and install `mason`.

File: tools/mason/MasonHelp.chpl [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+proc masonNewHelp() {
+  writeln('Usage:');
+  writeln('    mason new [options] <project name>');
+  writeln();
+  writeln('Options:');
+  writeln('    -h, --help                   Display this message');
+  writeln('        --show                   Increase verbosity');
+  writeln('        --novcs                  Do not initialize a git repository');
```
Comment (ben-albrecht) [inline tools/mason/MasonHelp.chpl]: I think this would be more consistent with `chpl` command line flags:

```
     --no-vcs                 Do not initialize a git repository
```

File: tools/mason/MasonHelp.chpl [hunk 30]
```diff
@@ -0,0 +1,82 @@
...
+proc masonBuildHelp() {
+  writeln('Compile a local package and all of its dependencies');
+  writeln();
+  writeln('Usage:');
+  writeln('    mason build [options]');
+  writeln();
+  writeln('Options:');
+  writeln('    -h, --help                   Display this message');
+  writeln('        --show                   Show compilation message');
+  writeln('        --release                Use --fast; Compile to target/release');
```
Comment (ben-albrecht) [inline tools/mason/MasonHelp.chpl]: I would reword this as `Compile to target/release with optimizations (--fast)` Maybe there will be other flags added to this in the future.

File: tools/mason/MasonHelp.chpl [hunk 31]
```diff
@@ -0,0 +1,82 @@
...
+
+proc masonBuildHelp() {
+  writeln('Compile a local package and all of its dependencies');
+  writeln();
+  writeln('Usage:');
+  writeln('    mason build [options]');
+  writeln();
+  writeln('Options:');
+  writeln('    -h, --help                   Display this message');
+  writeln('        --show                   Show compilation message');
```
Comment (ben-albrecht) [inline tools/mason/MasonHelp.chpl]: Let's stay consistent with the `Increase verbosity` language

File: tools/mason/MasonNew.chpl [hunk 32]
```diff
@@ -0,0 +1,98 @@
...
+  var tomlFile = open(name+"/Mason.toml", iomode.cw);
+  var tomlWriter = tomlFile.writer();
+  tomlWriter.write(baseToml);
+  tomlWriter.close();
+}
+  
+
+proc makeProjectFiles(name: string) {
+  mkdir(name + "/src");
+  const libTemplate = '\n /* Documentation for ' + name +
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl]: We can drop the leading `\n`

File: tools/mason/MasonNew.chpl [hunk 33]
```diff
@@ -0,0 +1,98 @@
...
+    ' */\nmodule '+ name + ' { \n   writeln("New library: '+ name +'");\n}';
```
Comment (ben-albrecht) [inline tools/mason/MasonNew.chpl]: There's a trailing space after `{`

File: tools/mason/MasonBuild.chpl [hunk 34]
```diff
@@ -0,0 +1,174 @@
...
+    }
+
+    // compile Program with deps
+    var compilation = runWithStatus(command);
+    if compilation != 0 {
+      return false;
+    }
+    
+    // Confirming File Structure
+    rename(project, 'target/' + binLoc + '/' + project);
```
Comment (ben-albrecht) [inline tools/mason/MasonBuild.chpl]: Could this be accomplished with `-o` flag in the compile command? This would also make `--show` output more reproducible for users.

File: doc/rst/tools/mason/mason.rst [hunk 35]
```diff
@@ -0,0 +1,339 @@
...
+
+
+
+
+
+
+Building and Running your project
+=================================
+
+When invoked, ``mason build`` will do the following:
```
Comment (ben-albrecht) [inline doc/rst/tools/mason/mason.rst]: We should mention how that `mason` forwards compile/runtime flags somewhere around here...

Also, I discovered an ugly caveat: 

Because `mason` is a Chapel program, it consumes all the default flags like `-v`, `-s`, `-nl`... This is annoying and problematic. Long-term, we'd like to create a way to disable these default flags for a Chapel program to disable this. In the near-term, this can be overridden by passing flags after a `--`.

So to build a Chapel program with a param flag via `mason`, you'd need to do:

```
> mason build -- -smyParamFlag=true
```

whereas the following will not work:

```
> mason build -smyParamFlag=true
<command-line arg>:2: error: Unrecognized configuration variable 'myParamFlag'
```

File: tools/mason/Makefile (hunk 27 above)
Comment (Spartee) [inline tools/mason/Makefile]: `make clean` should not(as cargo's does not), but I would like `make clobber` too.

Comment (ben-albrecht) [inline tools/mason/Makefile]: Also, `make uninstall` might be useful as well (only removes the symlinks in `$CHPL_HOME/bin/*/` -- this can be a future TODO, however.

File: tools/mason/MasonBuild.chpl [hunk 36]
```diff
@@ -0,0 +1,174 @@
...
+
+
+/* Creates the rest of the project structure */
+proc makeTargetFiles(binLoc: string) {
+  if !isDir('target') {
+    mkdir('target'); 
+  }
+  if !isDir('target/' + binLoc) {
+    mkdir('target/' + binLoc);
+    mkdir('target/' + binLoc + '/tests');
```
Comment (ben-albrecht) [inline tools/mason/MasonBuild.chpl]: What do you think about commenting these out, and adding these directories in as they become supported?

File: tools/mason/MasonBuild.chpl [hunk 37]
```diff
@@ -0,0 +1,174 @@
...
+  var release = false;
+  var compopts: [1..0] string;
+  if args.size > 2 {
+    for arg in args[2..] {
+      if arg == '-h' || arg == '--help' {
+	masonBuildHelp();
+	exit();
+      }
+      else if arg == '--release' {
+	release = true;
```
Comment (ben-albrecht) [inline tools/mason/MasonBuild.chpl]: Will we also support `--debug`, `--test`, `--bench`, etc. ?  What is the behavior when multiple modes are thrown?

File: tools/mason/MasonBuild.chpl [hunk 38]
```diff
@@ -0,0 +1,174 @@
...
+  BuildProgram(release, show, compopts);
+}
+
+
+proc BuildProgram(release: bool, show: bool, compopts: [?d] string) {
+  if isFile("Mason.lock") {
+
+    // --fast
+    var binLoc = 'debug';
+    if release {
```
Comment (ben-albrecht) [inline tools/mason/MasonBuild.chpl]: For 1-line `if-then` statements, you can do:

```chpl
if release then binLoc = 'release';
```

or with a newline (as some prefer):

```chpl
if release then 
  binLoc = 'release';
```

File: tools/mason/MasonBuild.chpl [hunk 39]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+
+/* Generates a list of tuples that holds the git repo
+   url and the name for local mason dependency pool */
+proc genSourceList(lockFile: Toml) {
+  var sourceList: [1..0] (string, string);
+  for (name, package) in zip(lockFile.D, lockFile.A) {
+    if package.tag == fieldToml {
+      if name == "root" {
```
Comment (ben-albrecht) [inline tools/mason/MasonBuild.chpl]:
```chpl
if name == "root" then continue;
```

File: tools/mason/MasonBuild.chpl [hunk 40]
```diff
@@ -0,0 +1,174 @@
...
+    if release {
+      command += " --fast";
+    }
+    for dep in sourceList {
+      var depSrc = ' -M '+ depPath + dep(2) + '/src';
+      command += depSrc;
+    }
+
+    // Verbosity control
+    if show {
```
Comment (ben-albrecht) [inline tools/mason/MasonBuild.chpl]:
```chpl
if show then writeln(command);
else writeln("Compiling ", project);

File: tools/mason/MasonUpdate.chpl [hunk 41]
```diff
@@ -0,0 +1,205 @@
...
+proc retrieveDep(name: string, version: string) {
+  var home = getMasonHome();
+  var tomlPath = home + "/.mason/registry/Bricks/"+name+"/"+version+".toml";
+  if isFile(tomlPath) {
+    var tomlFile = open(tomlPath, iomode.r);
+    var depToml = parseToml(tomlFile);
+    return depToml;
+  }
+  else {
+    halt("Could not find toml file for " + name);
```
Comment (ben-albrecht) [inline tools/mason/MasonUpdate.chpl]: This error pops up when providing an invalid version number. I think it'd be helpful to list the version as well. If we could distinguish the errors, that'd be even better (but also could be a next step), e.g.

```
# Wrong version number
Version for PackageFoo does not exist in registry
```

```
# Wrong package name
PackageFoo does not exist in registry
```

**Comment by ben-albrecht** [comment]
post-PR TODO: It would be nice to print clean error messages and `exit(1)` for errors rather than halting. A `halt()` prints a file and line number (and stacktrace if `CHPL_UNWIND`  is set), which comes off as an unhandled error more than an error message intended user consumption.

**Comment by ben-albrecht** [comment]
Some error-cases:

- Omitting `version` field in `[brick]` still compiles and runs (should it?)
- Omitting `name` field in `[brick]` results in a `dereference nil` on this line in `compileSrc()`:

```
  var project = lockFile["root"]["name"].s;
```

presumably, because `lockFile["root"]["name"]` is `nil`. Perhaps we could check for that earlier on...

I'd be OK handling these edge-cases in the next PR, but want to be sure to document them.

File: tools/mason/MasonBuild.chpl [hunk 42]
```diff
@@ -0,0 +1,174 @@
...
+   url and the name for local mason dependency pool */
+proc genSourceList(lockFile: Toml) {
+  var sourceList: [1..0] (string, string);
+  for (name, package) in zip(lockFile.D, lockFile.A) {
+    if package.tag == fieldToml {
+      if name == "root" {
+        continue;
+      }
+      else {
+        var version = lockFile[name]["version"].s;
```
Comment (ben-albrecht) [inline tools/mason/MasonBuild.chpl]: It's a good practice to make these `const` if we don't expect to modify them ever again.

File: tools/mason/MasonUtils.chpl [hunk 43]
```diff
@@ -0,0 +1,71 @@
...
+  for dir in listdir(repos) {
+    if dir == dependency {
+      exists = true;
+    }
+  }
+  return exists;
+}
+
+
+proc getMasonHome(): string {
```
Comment (ben-albrecht) [inline tools/mason/MasonUtils.chpl]: This is fine as is, but you could make this a parentheses-less function so that it appears to be a variable, e.g.

```chpl
proc MASON_HOME: string {
  ...
}
```

and calling it looks like:

```chpl
var depPath = MASON_HOME + '/.mason/src/';
```

**Comment by Spartee** [comment]
@ben-albrecht I think it is good to go! I took out one of the tests as I realized that it would fail and create a lot of noise. Will open a pr with a larger testing suite this week.

File: doc/rst/tools/mason/mason.rst [hunk 44]
```diff
@@ -5,18 +5,15 @@
...
 
 
 
 Installation Instructions
 =========================
 
-In the mason directory found in ``$CHPL_HOME/tools/mason`` run the following:
+In ``$CHPL_HOME`` run the following:
 
 .. code-block:: sh
```
Comment (ben-albrecht) [inline doc/rst/tools/mason/mason.rst]: We should add the target to the top-level `Makefile`.

File: doc/rst/tools/mason/mason.rst [hunk 45]
```diff
@@ -72,28 +60,26 @@ following hierarchy::
 
 Mason will enforce that the main file be named after the package to enforce namespacing.
 While it is common practice for package names to be PascalCase and chpl files to be lowercase,
-it is an acceptable tradeoff for reliability, and MyPackage will be the first file listed in ``src/``.
-
-
+it is an acceptable tradeoff for reliability. MyPackage will be the first file listed in ``src/``.
```
Comment (ben-albrecht) [inline doc/rst/tools/mason/mason.rst]: ``` ``MyPackage `` ```

Comment (Spartee) [inline doc/rst/tools/mason/mason.rst]: @ben-albrecht we dont do that anywhere else we say MyPackage so should that be a global change?

File: doc/rst/tools/mason/mason.rst [hunk 46]
```diff
@@ -161,9 +161,9 @@ The registry would follow a hierarchy as follows:
       2.2.0.toml
       2.2.1.toml
 
-Each versioned manifest file would be identical to the manifest file in the top-level directory
-of the package repository, with one additional field not required in the repository manifest,
-a URL pointing to the repository and revision in which the version is located.
+Each versioned manifest file is identical to the manifest file in the top-level directory
+of the package repository, with one exeption, a URL pointing to the repository and revision
```
Comment (ben-albrecht) [inline doc/rst/tools/mason/mason.rst]: exception

File: doc/rst/tools/mason/mason.rst [hunk 47]
```diff
@@ -127,9 +127,9 @@ For example, ``Mason.toml``:
 TOML
 ====
 
-TOML is the configuration language chosen by the chapel developers for
-configuring programs written in chapel using mason. A TOML file contains
-the nessescary information to build a chapel program using mason. 
+TOML is the configuration language chosen by the chapel team for
+configuring programs written in chapel. A TOML file contains the
+nessescary information to build a chapel program using mason.
```
Comment (ben-albrecht) [inline doc/rst/tools/mason/mason.rst]: necessary

File: tools/mason/MasonUtils.chpl [hunk 48]
```diff
@@ -3,7 +3,7 @@
 use Spawn;
 use FileSystem;
 
-/* Gets envoirnment variables for spawn commands */
+/* Gets enviornment variables for spawn commands */
```
Comment (ben-albrecht) [inline tools/mason/MasonUtils.chpl]: environment

File: tools/mason/Makefile [hunk 49]
```diff
@@ -0,0 +1,67 @@
...
+
+all: mason
+
+
+mason:	
+	@if [ $(shell $(CHPL_MAKE_HOME)/util/chplenv/chpl_regexp.py) != re2 ]; then \
+        echo "CHPL_REGEXP=re2 is required to build mason"; \
+        exit 1;\
+	fi;
+	@mkdir -p $(MASON_HOME)/.mason
```
Comment (ben-albrecht) [inline tools/mason/Makefile]: This line can be removed
